### ISSUE TITLE
NLU correctly extracts intent, does not pass to slot via story

### ISSUE THREAD

**Issue by Zylatis** [RasaHQ/rasa#3279]
**Rasa version**: 
Core: 0.13.7
Core-sdk: 0.13.0
NLU: 0.14.6

**Python version**:  3.6

**Operating system**:  Ubuntu 16.04

**Issue**: NLU correctly identifying intent, not passing to slot via story

**Content of configuration file (config.yml)**:
```yml
language: "en"
pipeline:
- name: "nlp_spacy"
- name: "tokenizer_spacy"
- name: "intent_entity_featurizer_regex"
- name: "intent_featurizer_spacy"
- name: "ner_crf"
- name: "intent_featurizer_count_vectors"
- name: "intent_classifier_tensorflow_embedding"
- name: "ner_duckling_http"
  url:  "http://0.0.0.0:8000"
  dimensions: ["time"]
```

**Content of domain file (domain.yml)** (if used & relevant):
```yaml
intents:
  - greet
  - goodbye
  - get_quote_withinfo
  - get_quote
  - modify_booking
  - thanks
  - inform_city
  - inform_date
  - check_status

actions:
- utter_greet
- utter_goodbye
- utter_welcome
- utter_get_quote
- utter_slot_values
- utter_submit
- utter_ask_source # should this not be included by default, will i be overriding shit?
- utter_changebook
- utter_check_status

entities:
  - source
  - destination
  - day
  - month
  - year
  - time_day
  - new_day
  - new_month
  - date
  - city
  - time

slots:
  source:
    type: unfeaturized
    auto_fill: false
  destination:
    type: unfeaturized
    auto_fill: false
  date:
    type: unfeaturized
    auto_fill: false
  day:
    type: unfeaturized
    auto_fill: false
  month:
    type: unfeaturized
    auto_fill: false
  year:
    type: unfeaturized
    auto_fill: false
  time_day:
    type: unfeaturized
    auto_fill: false

templates:
  utter_greet:
  - text: "Hello! How can I help?"
  - text: "Hiya!"
  utter_goodbye:
  - text: "Goodbye!"
  utter_ask_source:
    - text: "Where are you shipping from?"
  utter_ask_destination:
    - text: "Where are you shipping to?"
  utter_ask_date:
  - text: "When do you want your items to ship from {source}?"
  utter_welcome:
    - text: "You're welcome!"
    - text: "No worries :)"
  utter_get_quote:
    - text: "Okay, lets get a quote going"
  utter_changebook:
    - text: "Okay, let me grab that booking for you"
  utter_slot_values:
    - text: "I am going to run a quote search using the following parameters:\n
            - source: {source}\n
            - destination: {destination}\n
            - date: {day} of {month}\n
            - time: {time_day}"
  utter_submit:
    - text: "All done!"
  utter_default:
    - text: "Totes soz, I don't understand what you mean there, can you rephrase please?"
  utter_ask_month:
    - text: "What day of {year}"
  utter_ask_day:
    - text: "What day of {month}"
  utter_ask_time_day:
    - text: "What time on {day}"

forms:
  - booking_form
```

**Content of stories.md
```md
## new_quote2
* greet
  - utter_greet
* get_quote{"source":"perth"}
  - utter_get_quote
  - booking_form
  - form{"name": "booking_form"}
  - form{"name": null}
  - utter_slot_values
* goodbye
  - utter_goodbye
```

I am trying to make a fairly standard variation of the form bot which will make bookings. If the user writes 'make booking' and starts the form then everything works fine. However, I would like to be able to extract information from the very first sentence before an intent has been classified (i.e. before a form is used). It seems this is done via slot filling directly in the stories but i cannot get it to work. In the attached image it shows that NLU is correctly pulling the source city (though admittedly low confidence but that's fine) but the form is still asking for it.

 I have tried several variations (i.e. putting the slot filling part in the story after the form is activated) but nothing seems to work, what am i missing?

Thanks!
![nluform](https://user-images.githubusercontent.com/1456631/56335676-c9ef9100-61e0-11e9-9968-d62bf48f7ee0.png)

**Comment by akelad** [comment]
Thanks for raising this issue, @Ghostvv  will get back to you about it soon.

**Comment by Ghostvv** [comment]
what version of `rasa_core_sdk` are you using? it didn't validate already filled slots. We updated forms to add this functionality: https://github.com/RasaHQ/rasa_core_sdk/pull/67

Could you please try a new version

**Comment by Zylatis** [comment]
Hello,

The versions I used for that example are listed at the top of the post. I did a pip install --upgrade and am now running 
rasa-core                     0.13.8                
rasa-core-sdk                 0.13.1                
rasa-nlu                      0.14.6 

but the issue persists. NLU correctly pulls the entity but the slot remains unfilled. Is there a particular version i should specifically request through pip? I have attached another image showing the same thing happening.

Thanks!
![eg](https://user-images.githubusercontent.com/1456631/56464381-28a84b00-642b-11e9-9554-1848111943bd.png)

**Comment by Zylatis** [comment]
Update: okay I *think* I have a better idea of my confusion, and I have made a more compact example to illustrate this.

Let's say I have two slots to fill, pickup_city and pickup_street. I have two entities of the same name, so I can extract them directly, but I also want to be able to extract them from a more general intent 'inform_address' which may be used for pickup but also for dropoff. So, for this intent i have two other entities just 'street' and 'city.

What I thought would happen is if i have the following form definition:

` @staticmethod
	 def required_slots(tracker: Tracker) -> List[Text]:
			 return ["pickup_city", "pickup_street"]
			
	 def slot_mappings(self):
				return {
					"pickup_city": [ self.from_entity(entity="pickup_city", intent = ["make_booking"]),self.from_entity(entity="city",intent=["inform"])],
					"pickup_street": [self.from_entity(entity="pickup_street", intent = ["make_booking"]),self.from_entity(entity="street", intent = ["inform"])]
				}`

that if the form asked for 'pickup_city' but I gave 'city' and 'street' (as correctly extracted by NLU), then the slot mapping would not only give me pickup_city but also see that it extracted 'street' and map that to 'pickup_street'. However, this does not happen: it does not see that it can fill a slot it knows about by using an entity it has just seen.

For example, if I change it so the required slot it 'street' then this works: I give it a city and a street at the same time and it sees 'street' as the extra slot to fill and does so (though here it would not work because i have not defined 'street' as a slot but i can see it is trying to do the right thing anyway)

So, while I understand i could get around this by having a custom 'validate' function, am I wrong in thinking this sort of thing should happen by default?

Thanks again!

**Comment by Ghostvv** [comment]
Do you have another action or utterance between a user input and a form? By default, validation is triggered only if the form is called after `action_listen`

**Comment by Zylatis** [comment]
I did have that, though if I remove it i still get the same thing. I have changed the setup quite a bit since the last round, sorry, so here is the new setup.

```
## intent:get_quote
- get quote
- make quote
- hello, i want to get a quote
- hi i need a quote
- hey i want to find out how much it will cost
- how much will it cost to send stuff from [bayswater](pickup_suburb) to [castlemaine](dropoff_suburb)?
- i want to send some items
- i want to ship some freight
- i want to ship items to [moonah](dropoff_suburb)
- i want to send items to [mildura](dropoff_suburb)
- i need to ship to [petersham](dropoff_suburb)
- i want to send some items from [brisbane](pickup_suburb)
- i want to ship some freight [rye](pickup_suburb)
- i want to send items from [petersham](pickup_suburb) to [brisbane](dropoff_suburb), how much will it cost?
- i need a quote on sending from [new town](pickup_suburb) to [ultimo](dropoff_suburb)
- i want to send items to [petersham](dropoff_suburb) from [brisbane](pickup_suburb), how much will it cost?
- i need a quote on sending to [new town](dropoff_suburb) from [ultimo](pickup_suburb)
- how much will it cost to send stuff to [bayswater](dropoff_suburb) from [castlemaine](pickup_suburb)?
```

```
## new_quote1
- utter_intro
* get_quote
  - pickup_address_form
  - form{"name": "pickup_address_form"}
  - form{"name": null}
  - dropoff_address_form
  - form{"name": "dropoff_address_form"}
  - form{"name": null}
  - items_form
  - form{"name": "items_form"}
  - form{"name": null}
  - utter_return_quote
```

```
 @staticmethod
 def required_slots(tracker: Tracker) -> List[Text]:
 	return ["pickup_suburb", "pickup_postcode", "pickup_residential"]
		
 def slot_mappings(self):
	return {
		"pickup_suburb": [ self.from_entity(entity="pickup_suburb", intent = ["make_booking"]),self.from_entity(entity="suburb",intent=["inform_address"])],
		"pickup_postcode": [self.from_entity(entity="pickup_postcode", intent = ["make_booking"]), self.from_entity(entity="postcode", intent = ["inform_address"]), self.from_entity(entity="number", intent = ["inform_address"])],
		"pickup_residential": [ self.from_intent(intent='affirm', value=True), self.from_intent(intent='deny', value=False)]
	}
 
 # NOTE: this is infact called for each slot at time of filling, not at the end
 def validate(self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict[Text, Any]) -> List[Dict]:
		
	slot_values = self.extract_other_slots(dispatcher, tracker, domain)
	slot_to_fill = tracker.get_slot(REQUESTED_SLOT)
	if slot_to_fill:
		slot_values.update(self.extract_requested_slot(dispatcher, tracker, domain))
		if not slot_values:
			raise ActionExecutionRejection(self.name(), "Failed to validate slot {0} "
														"with action {1}"
															"".format(slot_to_fill,self.name())
															)

	latest_message = tracker.latest_message['entities']
	updated_slots = {}

	for el in latest_message:
		value = el['value']
		if el['entity'] == 'postcode':
			updated_slots['pickup_postcode'] = value
		elif el['entity'] == 'suburb':
			updated_slots['pickup_suburb'] = value
			
			
	for slot, value in updated_slots.items():
		slot_values[slot] = value
	
	return [SlotSet(slot, value) for slot, value in slot_values.items()]
```

I have attached an image of the output. Is it possible because i have overwritten the validate() function it is no longer looking for other prefilled stuff? The Core output shows NLU extracting the **pickup_suburb** correctly but it still does not apply it to the appropriate slot
![new_core](https://user-images.githubusercontent.com/1456631/56628196-37da0380-668c-11e9-86a9-414044354ea8.png)

Thanks!

**Comment by Zylatis** [comment]
A thousand apologies, I had changed one of my intents and failed to update my actions.py

For now this seems to work swimmingly so I am happy to close the issue, however it would still be good to know how to utter a message at the start of a form without losing the slot filling aspect.

**Comment by Ghostvv** [comment]
there is `_activate_if_required` method, which you can copy and add utterance upon activation there.
there is slot mapping `from_trigger_intent` that should fill the slot upon form activation depending on user intent that triggered the form

**Comment by Ghostvv** [comment]
`however it would still be good to know how to utter a message at the start of a form without losing the slot filling aspect.` - we thinking about the most general way to implement it

**Comment by Zylatis** [comment]
Ah okay great, will play with that. For now I just have a custom action

```
class FillAll(Action):
	def name(self):
		return "action_fillall"

	def run(self, dispatcher, tracker, domain):
		latest_message = tracker.latest_message
		entities = latest_message['entities']
		slots = tracker.slots.keys()

		result = []
		for el in entities:
			if el['entity'] in slots:
				result.append(SlotSet(el['entity'], el['value']))
		return result
```

which has zero error handling and is lazy but seems to work.

Regarding 'best way' ideally, at least from myself as a user point of view, something akin to the submit() method, basically an init(), would be great.

**Comment by Zylatis** [comment]
Hello,

I'm afraid i need to re-open this issue. I went back to a previous version of forms/stories and have come up against this issue again, though it may be me expecting something that isn't meant to happen. Note I have re-read the above stuff to make sure i'm not re-hashing old solutions.

Problem is simple: given the following FormAction (leaving validate() un-overloaded)
 
```python
class QuoteForm(FormAction):
    
    def name(self):
        return "quote_form"

    @staticmethod
    def required_slots(tracker: Tracker) -> List[Text]:
        return ["pickup_suburb", "pickup_postcode"  ]
                
    def slot_mappings(self):
        return {
            "pickup_suburb": [ 
                self.from_entity(entity="pickup_suburb", intent = ["quote"]),
                self.from_entity(entity="suburb",intent=["inform_address"])],
            "pickup_postcode": [ 
                self.from_entity(entity="postcode", intent = ["inform_address"]),
                self.from_entity(entity="pickup_postcode", intent = ["quote"]),
                self.from_entity(entity="number", intent = ["inform_number"])]  
```

and story 
```md
## quote
* greet
  - utter_greet
* quote
  - quote_form
  - form{"name":"quote_form"}
  - form{"name": null}
```

I want to be able to respond to suburb and postcode at the same time when suburb is asked for. Specifically, when it asks for suburb  it actually asks for suburb and postcode, so the user has the option to enter both at the same time. I had assumed that the mappings defined above would mean that it would get the postcode even though only suburb was asked for, is that not the case?

An example response i am trying to parse when asked for suburb is " [perth](suburb) [2000](postcode)" with intent inform_address

 I can put a whole bunch of things in validate to deal with this, but it would involve lots of intent checking which should be handled already by the slot mappings (I will have pickup/dropoff and so on)

Thanks!
Z.

**Comment by Zylatis** [comment]
Update: Okay I think i have answered my own question, apologies for the threadnaught.

I managed to hack my way to what i wanted but turns out what i wanted isn't what i needed, as i end up overriding slots ahead of time. I think i saw mention of 'groups' being maybe implemented in the future which may solve this problem. 

For now I think I  will just define sets of slots that may be set together and others which can only be set when asked for.

Thanks!

**Comment by akelad** [comment]
so can we close this issue again?

**Comment by Zylatis** [comment]
Yes, my apologies, thanks!
